warehouse/accounts/services.py

```diff
@@ -63,6 +74,42 @@ def get_user(self, userid):
         # TODO: We need some sort of Anonymous User.
         return self.db.query(User).get(userid)
 
+    def get_user_by_otk(self, otk):
+        if not otk:
+            raise InvalidPasswordResetToken
+
+        try:
+            data = self.serializer.loads(
+                otk,
+                self.token_max_age
+            )
+        except BadData:
+            raise InvalidPasswordResetToken
+
+        # Check whether a user with the given user ID exists
+        user = self.get_user(uuid.UUID(data.get("user.id")))
+        if user is None:
+            raise InvalidPasswordResetToken
+
+        # Compare the otk hash values for the user
+        user_hash = self._generate_otk_hash(user)
+        if not hmac.compare_digest(data.get("user.hash"), user_hash):
+            raise InvalidPasswordResetToken
+
+        return user
+
+    @functools.lru_cache()
+    def get_user_by_username(self, username):
```

Comment:
If we're going to add a public method like this, it needs to be added to the ``IUserService`` interface as well. 

One thing to check, I believe the way SQLAlchemy works, you could implement this method as just doing:

```python
def get_user_by_username(self, username):
    user_id = self.find_userid(username)
    return None if user_id is None else self.get_user(user_id)
```

SQLAlchemy should still have the ``User`` class in it's cache and can just return that existing instance rather than re-querying the database (I think anyways!). That should make the implementation of this method a little more DRY. It also raises the question of if this method should be part of ``IUserService`` or if it makes sense to just use embed those two lines of code into the view. Neither answer is wrong here (and adding it to the ``IUserService`` does make it possible to optimize the queries if I'm wrong about SQLAlchemy or we switch to another system that doesn't do that).